core-rust/state-manager/src/store/rocks_db.rs

```diff
@@ -1842,15 +1828,21 @@ impl<R: WriteableRocks> StateTreeGcStore for StateManagerDatabase<R> {
         }
     }
 
-    fn batch_delete_stale_tree_part<'a>(
-        &self,
-        state_versions: impl IntoIterator<Item = &'a StateVersion>,
-    ) {
+    fn delete_stale_tree_parts_up_to_version(&self, state_version: StateVersion) {
         let db_context = self.open_rw_context();
-        let stale_tree_parts_cf = db_context.cf(StaleStateTreePartsCf);
-        for state_version in state_versions {
-            stale_tree_parts_cf.delete(state_version);
-        }
+        db_context
+            .cf(StaleStateTreePartsCf)
+            .delete_range(&StateVersion::pre_genesis(), &state_version);
+        let updated_status = StateTreeAssociatedValuesStatusV1 {
```

Comment:
It's a little weird to me that we combine the deletion of the stale tree notifications with the updating of the status.

In fact, I think this is invalid and would cause a panic, because we update `VersionedStateTreeAssociatedValuesStatus` too late after we've already deleted some substate values.

The updating of the status should be tied to the presence of the substate values - so it would need to happen in the `batch_delete_node` method - we should call a `invalidate_historical_substate_value_availability_up_to(max_version_deleted)` or something - and the `invalidate_historical_substate_value_availability_up_to` should load the current status, and take max of current status, and `invalid_up_to + 1`.

Reply:
OMG, nice catch - yeah, that would be a panic every time someone accesses historical state during JMT GC...

I've re-written this part, with more rustdocs and comments - PTAL.